packages/astro/src/assets/vite-plugin-assets.ts

```diff
@@ -0,0 +1,229 @@
+import MagicString from 'magic-string';
+import mime from 'mime';
+import fs from 'node:fs/promises';
+import path from 'node:path';
+import { Readable } from 'node:stream';
+import { fileURLToPath, pathToFileURL } from 'node:url';
+import type * as vite from 'vite';
+import { normalizePath } from 'vite';
+import { AstroPluginOptions, ImageTransform } from '../@types/astro';
+import { joinPaths, prependForwardSlash } from '../core/path.js';
+import { VIRTUAL_MODULE_ID, VIRTUAL_SERVICE_ID } from './consts.js';
+import { isESMImportedImage } from './internal.js';
+import { isLocalService } from './services/service.js';
+import { copyWasmFiles } from './services/vendor/squoosh/copy-wasm.js';
+import { imageMetadata } from './utils/metadata.js';
+import { getOrigQueryParams } from './utils/queryParams.js';
+import { propsToFilename } from './utils/transformToPath.js';
+
+const resolvedVirtualModuleId = '\0' + VIRTUAL_MODULE_ID;
+
+export default function assets({
+	settings,
+	logging,
+	mode,
+}: AstroPluginOptions & { mode: string }): vite.Plugin[] {
+	let resolvedConfig: vite.ResolvedConfig;
+
+	globalThis.astroAsset = {};
```

Comment:
In the meantime, what do you think about tucking away the use of this global into its own module. So the nastiness can be contained to within there? Something like:

```js
export function createGlobal() {
  globalThis.astroAsset = {};
}

export default function() {
  return globalThis.astroAsset;
}

export function destroyGlobal() {
  delete globalThis.astroAsset;
}
```

Then, here you would do:

```js
import { createGlobal } from './asset.js';


....
createGlobal();
```

And anywhere that just needs to access it does:

```js
import astroAsset from '/path/to/assets/asset.js';

// Get the image service
astroAsset().imageService; // ....

// Call the addStaticImage function
astroAsset().addStaticImage(/* ... */);
```

etc. Then when we refactor to remove the global hopefully anything using it wouldn't need to change.

Just an idea, let me know if you don't think it'll help enough.

Reply:
I tried to do this, because it seemed like a good idea, but unfortunately TypeScript made this hell and it lead to needing to add a bunch of unnecessary checks everywhere 😦

Reply:
ok, we'll revisit later.